src/hotp/hotp.c

```diff
@@ -493,11 +493,13 @@ uint8_t* page = (uint8_t *) current_slot_address;
 
     memcpy (page_buffer, page, SLOT_PAGE_SIZE);
 
-    // check if the secret from the tool is empty and if it is use the old
-    // secret
-uint8_t* secret = (uint8_t *) (data + SECRET_OFFSET);
-
-    if (secret[0] == 0)
+    // Check if the secret from the tool is empty and if it is use the old secret
+    // Secret could begin with 0x00, so checking the whole secret before keeping the old one in mandatory
+    // The while last loop points secret[-1]. Well, it's right, == result does not matter and mem exits, we need a -1
+    uint8_t* secret = (uint8_t *) (data + SECRET_OFFSET);
+    int8_t i = 20;
+    while (i >= 0 && secret[--i] == 0x00);
```

Comment:
@MaVo159 , I'm agree, your proposal is clear and less error-prone. However, I'm not used to be verbose for fixes like this, trying to do the less mods as possible, also the protocol is simple, lacks a size value giving us the possibility of write simpler code  `if (secret_size == 0)` , so on I decided to do it in this way. Feel free to change.

As @szszszsz said, the code is safe any way, position `-1` is never accessed.